synthesizer/src/vm/mod.rs

```diff
@@ -332,15 +332,30 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             block.previous_hash(),
         )?;
 
-        // Attention: The following order is crucial because if 'finalize' fails, we can rollback the block.
-        // If one first calls 'finalize', then calls 'insert(block)' and it fails, there is no way to rollback 'finalize'.
+        // Enable the atomic batch override, so that both the insertion and finalization belong to a single batch.
+        #[cfg(feature = "rocks")]
+        assert!(self.block_store().flip_atomic_override()?);
 
         // First, insert the block.
         self.block_store().insert(block)?;
         // Next, finalize the transactions.
         match self.finalize(state, block.ratifications(), block.solutions(), block.transactions()) {
-            Ok(_ratified_finalize_operations) => Ok(()),
+            Ok(_ratified_finalize_operations) => {
+                // Disable the atomic batch override, executing it.
+                #[cfg(feature = "rocks")]
+                assert!(!self.block_store().flip_atomic_override()?);
+                Ok(())
+            }
             Err(finalize_error) => {
+                #[cfg(feature = "rocks")]
+                {
+                    // Clear all pending atomic operations so that disabling the atomic batch override
+                    // doesn't execute any storage operation that accumulated after it was enabled.
+                    self.block_store().atomic_abort();
+                    self.finalize_store().atomic_abort();
+                    // Disable the atomic batch override.
+                    assert!(!self.block_store().flip_atomic_override()?);
+                }
                 // Rollback the block.
                 self.block_store().remove_last_n(1).map_err(|removal_error| {
```

Comment:
This is still incorrect.

Calling `remove_last_n(1)` here will 1) correct update the Merkle tree, however 2) it will also delete the penultimate block now from `self.block_store()`.

This is erroneous because we are using this concept of a `flip_atomic_override`.

Reply:
Indeed; there are 2 ways I can see of tackling this; we can either:
- put the storage deletion within `remove_last_n` behind a feature flag (so that it doesn't happen with `rocks`); however, this would obscure the purpose of the method and affect its use outside of typical runtime (and probably break some tests)
- break the `remove_last_n` method into 2 distinct operations (isolated into new methods) - the tree update and the storage update - and only call the tree update here; I'll see if I can propose something "clean" for this

Reply:
I proposed the 2nd approach in https://github.com/AleoHQ/snarkVM/pull/2296/commits/1a27305b7e0acb2c23874bdf1d11ce2b527746e1.